### ISSUE TITLE
Setup automatic deployment to staging site

### ISSUE THREAD

**Issue by vitorbaptista** [opentrials/api#23]
After #17, Travis builds a Docker image and pushes to `opentrials/api:latest` on the master branch. This is fine for production, but as we develop inside branches, it's useless for staging. I'd like a way to test the current state of the code before it's merged into master.

My idea is to deploy branches other than `master` to `opentrials/api:staging`. This way, we can setup docker cloud to monitor this tag for changes.

There's an obvious problem: if there're multiple branches being worked on at a specific time, how to know which one is in staging? I can't see an easy solution: if we have a single staging environment, there can be only one branch deployed. Considering that the dev team for this project is very small, I think we can live with it for now. Staging will have the latest branch that was pushed to GitHub (which isn't master).

**Comment by pwalsh** [comment]
An alternate solution, similar to our continuous deployment instructions in coding standards, would be to always push master to staging, and only push tagged commits to production

**Comment by roll** [comment]
Don't forget I've already found one of solutions - https://github.com/opentrials/opentrials/issues/31 - just don't have a time to implement somewhere.

For now for `collectors` and `processors` I use deploy by commit message:
``` 
some commit [deploy]
```
goes for production (because there is no staging). It's less annoying than tags. Could be extended to `[production]`, `[staging]` (all always from master) etc marks.

**Comment by vitorbaptista** [comment]
@roll I've just pushed the commit https://github.com/opentrials/api/commit/349a0a8b84e22b61d138b3efa0b4a127bbe75075 which implements your idea described on https://github.com/opentrials/opentrials/issues/31#issuecomment-197188322.

Unfortunately, it ended up being quite complex. The main challenges are:

1. Setup the environment variables correctly based on if we're on staging or production;
2. Use different stackfiles depending on the environment, not only because of configuration differences (e.g. in staging we use the database in a container as well, while in production we don't), but because we're using different docker images (staging uses `opentrials/api:staging` while production uses `opentrials/api:latest`).
3. Create a new stack of update an existing one, depending if there's an existing one or not (docker-cloud doesn't have this "upsert" functionality by default).

It seems that Docker Cloud doesn't support extending stackfiles like [Docker Compose does](https://docs.docker.com/compose/extends/#extending-services), so there's quite a bit of repetition.

We also need to set different ports depending on the environment if we're deploying them to the same node (or leave them as random), because there can't be multiple containers listening on the same port.

It's quite late, so I couldn't test the deployment process going through Travis. It's working locally, though.

What do you think of the current code? I tried to code defensively checking everything along the way, and should've written it in Python instead of Bash. However, I'm afraid it'll be a burden to maintain, as there're many moving parts.

**Comment by roll** [comment]
Hi, my thoughts.

#### Image
Why do we need different images? For me image is just a reflection (builded) of code repository. Then image usage could be altered two ways:
  - by docker command
    - run server
    - run migrations for example 
  - by environment variables
    - staging
    - production

#### Stacks

I think our goal should be to have the same image for `stating` and `production` because it will lead for testing exactly the same infrastructure which will be in production. I think it's one of the primary motivations to use containerized infrastructure. @pwalsh could we have staging db? Or may be use different postgres schema?

#### Scripts
My choice to do not be so generous - it's not an public API - just internal script like:
```
deploy:
- provider: script
  on: <staging>
  script: staging.sh
```
```scripts/staging.sh
export DATABASE_USE=DATABASE_URL__STAGING 
bash deploy.sh
```
```scripts/deploy.sh
docker-cloud ...
...
```

**Comment by vitorbaptista** [comment]
We need different images because they're running different code. On `production` we have what's on master, but on `staging` we have the code of some other branch.

**Comment by pwalsh** [comment]
I'm closing this. we are only using Docker Cloud for the warehouse for the foreseeable future.
